In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score

df = pd.read_csv(r'C:\Users\admin\superstore-sales-analysis\data\superstore.csv\superstore.csv', encoding='utf-8')
df = df.rename(columns={'记录数': 'Record.Count'})

df.shape

(51290, 27)

In [4]:
X = df[['Sales', 'Discount']]
y = df['Profit']

print(X.head())
print(y.head())

   Sales  Discount
0     19       0.0
1     19       0.0
2     21       0.0
3    111       0.0
4      6       0.0
0     9.3312
1     9.2928
2     9.8418
3    53.2608
4     3.1104
Name: Profit, dtype: float64


In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("حجم بيانات التدريب:", X_train.shape)
print("حجم بيانات الاختبار:", X_test.shape)

حجم بيانات التدريب: (41032, 2)
حجم بيانات الاختبار: (10258, 2)


In [6]:
model = LinearRegression()
model.fit(X_train, y_train)

print("تم تدريب النموذج بنجاح")

تم تدريب النموذج بنجاح


In [7]:
y_pred = model.predict(X_test)

# مقارنة أول 10 قيم حقيقية بالقيم المتوقعة
comparison = pd.DataFrame({'Actual': y_test.values[:10], 'Predicted': y_pred[:10]})
print(comparison)

    Actual   Predicted
0   5.0328  -25.004134
1  37.2120   22.964175
2   8.0400   19.197487
3 -11.1000  -87.948328
4  -2.0988  -17.924608
5   0.1500   17.382224
6   2.5200   19.742066
7  -8.4270 -136.506581
8   3.5520  -65.575228
9 -29.1600  -58.858755


In [8]:
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("متوسط الخطأ المطلق (MAE):", mae)
print("R² Score:", r2)

متوسط الخطأ المطلق (MAE): 57.550571387803494
R² Score: 0.17039923962310877


In [9]:
X2 = df[['Sales', 'Discount', 'Shipping.Cost']]
y2 = df['Profit']

X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size=0.2, random_state=42)

model2 = LinearRegression()
model2.fit(X_train2, y_train2)

y_pred2 = model2.predict(X_test2)

mae2 = mean_absolute_error(y_test2, y_pred2)
r22 = r2_score(y_test2, y_pred2)

print("MAE الجديد:", mae2)
print("R² الجديد:", r22)

MAE الجديد: 58.04349038790649
R² الجديد: 0.13525779065845245


In [10]:
# تحويل عمود Category النصي لأعمدة رقمية (One-Hot Encoding)
df_encoded = pd.get_dummies(df, columns=['Category'], drop_first=True)

# شوف الأعمدة الجديدة اللي اتعملت
print([col for col in df_encoded.columns if 'Category' in col])

['Sub.Category', 'Category_Office Supplies', 'Category_Technology']


In [11]:
X3 = df_encoded[['Sales', 'Discount', 'Category_Office Supplies', 'Category_Technology']]
y3 = df_encoded['Profit']

X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y3, test_size=0.2, random_state=42)

model3 = LinearRegression()
model3.fit(X_train3, y_train3)

y_pred3 = model3.predict(X_test3)

mae3 = mean_absolute_error(y_test3, y_pred3)
r23 = r2_score(y_test3, y_pred3)

print("MAE مع الفئة:", mae3)
print("R² مع الفئة:", r23)

MAE مع الفئة: 59.31144796811649
R² مع الفئة: 0.16579213047747654


In [12]:
from sklearn.tree import DecisionTreeRegressor

model4 = DecisionTreeRegressor(max_depth=5, random_state=42)
model4.fit(X_train3, y_train3)

y_pred4 = model4.predict(X_test3)

mae4 = mean_absolute_error(y_test3, y_pred4)
r24 = r2_score(y_test3, y_pred4)

print("MAE مع Decision Tree:", mae4)
print("R² مع Decision Tree:", r24)

MAE مع Decision Tree: 40.61061281016265
R² مع Decision Tree: 0.568383373110619


In [13]:
from sklearn.ensemble import RandomForestRegressor

model5 = RandomForestRegressor(n_estimators=100, max_depth=8, random_state=42)
model5.fit(X_train3, y_train3)

y_pred5 = model5.predict(X_test3)

mae5 = mean_absolute_error(y_test3, y_pred5)
r25 = r2_score(y_test3, y_pred5)

print("MAE مع Random Forest:", mae5)
print("R² مع Random Forest:", r25)

MAE مع Random Forest: 35.87006809460801
R² مع Random Forest: 0.6325714589270741


In [14]:
# نعمل Encoding لعمود Sub.Category كمان
df_encoded2 = pd.get_dummies(df, columns=['Category', 'Sub.Category'], drop_first=True)

# نجهز المتغيرات الجديدة (كل أعمدة الفئة والفئة الفرعية + Sales + Discount)
subcategory_cols = [col for col in df_encoded2.columns if 'Sub.Category' in col or 'Category_' in col]
feature_cols = ['Sales', 'Discount'] + subcategory_cols

X4 = df_encoded2[feature_cols]
y4 = df_encoded2['Profit']

X_train4, X_test4, y_train4, y_test4 = train_test_split(X4, y4, test_size=0.2, random_state=42)

model6 = RandomForestRegressor(n_estimators=100, max_depth=8, random_state=42)
model6.fit(X_train4, y_train4)

y_pred6 = model6.predict(X_test4)

mae6 = mean_absolute_error(y_test4, y_pred6)
r26 = r2_score(y_test4, y_pred6)

print("MAE مع Sub.Category:", mae6)
print("R² مع Sub.Category:", r26)

MAE مع Sub.Category: 35.56674596577493
R² مع Sub.Category: 0.6336242361206159


# Machine Learning - Profit Prediction Model

## Objective
Build a regression model to predict Profit based on Sales, Discount, and Category features.

## Model Comparison

| Model | Features | R² Score | MAE |
|---|---|---|---|
| Linear Regression | Sales, Discount | 0.170 | 57.55 |
| Linear Regression | + Shipping.Cost | 0.135 | 58.04 |
| Linear Regression | + Category | 0.166 | 59.31 |
| Decision Tree | Sales, Discount, Category | 0.568 | 40.61 |
| Random Forest | Sales, Discount, Category | 0.633 | 35.87 |
| Random Forest | + Sub.Category | 0.634 | 35.57 |

## Key Findings
1. The relationship between Discount and Profit is **non-linear** — linear models performed poorly (R² < 0.2)
2. Tree-based models (Decision Tree, Random Forest) captured the non-linear discount threshold effect much better
3. Random Forest achieved the best performance (R² = 0.63), improving over a single Decision Tree by reducing overfitting
4. Adding Sub.Category provided minimal improvement, as its information overlaps with Category (diminishing returns)

## Final Model
**Random Forest Regressor** (n_estimators=100, max_depth=8)
- R² Score: 0.633
- MAE: $35.87

## Tools Used
Python, Pandas, Scikit-learn (LinearRegression, DecisionTreeRegressor, RandomForestRegressor)